File: java/base/src/main/vsdk/toolkit/common/linealAlgebra/Quaternionf.java
package vsdk.toolkit.common.linealAlgebra;

import java.io.Serial;
import java.util.Objects;

import vsdk.toolkit.common.FundamentalEntity;
import vsdk.toolkit.common.VSDK;

public final class Quaternionf extends FundamentalEntity
{
    @Serial
    private static final long serialVersionUID = 20260514L;

    private final Vector3Df direction;
    private final float magnitude;

    public Quaternionf()
    {
        this(new Vector3Df(0f, 0f, 0f), 0f);
    }

    public Quaternionf(Vector3Df direction, float magnitude)
    {
        this.direction = new Vector3Df(Objects.requireNonNull(direction, "Quaternion direction cannot be null"));
        this.magnitude = magnitude;
    }

    public Quaternionf(Quaternionf other)
    {
        this(Objects.requireNonNull(other, "Quaternion to copy cannot be null").direction,
             other.magnitude);
    }

    public Quaternionf(Quaterniond other)
    {
        this(new Vector3Df(other.direction()), (float) other.magnitude());
    }

    public static Quaternionf copyOf(Quaternionf other)
    {
        return new Quaternionf(Objects.requireNonNull(other, "Quaternion to copy cannot be null"));
    }

    public float lengthSquared()
    {
        return magnitude * magnitude + direction.dotProduct(direction);
    }

    public float length()
    {
        return (float) Math.sqrt(lengthSquared());
    }

    public Quaternionf normalized()
    {
        float l = length();
        if ( Math.abs(l) < (float) VSDK.EPSILON ) {
            return this;
        }
        float inv = 1f / l;
        return new Quaternionf(direction.multiply(inv), magnitude * inv);
    }

    public Quaternionf conjugated()
    {
        return new Quaternionf(direction.multiply(-1f), magnitude);
    }

    public Vector3Df rotate(Vector3Df vector)
    {
        Vector3Df normalizedAxis = direction;
        Vector3Df uv = normalizedAxis.crossProduct(
            Objects.requireNonNull(vector, "Vector to rotate cannot be null"));
        Vector3Df uuv = normalizedAxis.crossProduct(uv);

        return vector.add(uv.multiply(2f * magnitude)).add(uuv.multiply(2f));
    }

    public Quaternionf withDirection(Vector3Df newDirection)
    {
        return new Quaternionf(newDirection, magnitude);
    }

    public Quaternionf withMagnitude(float newMagnitude)
    {
        return new Quaternionf(direction, newMagnitude);
    }

    public Vector3Df direction()
    {
        return direction;
    }

    public float magnitude()
    {
        return magnitude;
    }

    public boolean epsilonEquals(Quaternionf other)
    {
        return epsilonEquals(other, (float) VSDK.EPSILON);
    }

    public boolean epsilonEquals(Quaternionf other, float epsilon)
    {
        if ( other == null ) {
            return false;
        }
        if ( epsilon < 0.0f ) {
            throw new IllegalArgumentException("epsilon must be >= 0");
        }
        return direction.epsilonEquals(other.direction, epsilon) &&
               Math.abs(magnitude - other.magnitude) <= epsilon;
    }

    @Override
    public String toString()
    {
        return direction + " / " + VSDK.formatDouble(magnitude);
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/linealAlgebra/Quaternionf.ts
import { FundamentalEntity } from "../FundamentalEntity.js";
import { VSDK } from "../VSDK.js";
import { Vector3Df } from "./Vector3Df.js";
import { Quaterniond } from "./Quaterniond.js";
const f = Math.fround;
export class Quaternionf extends FundamentalEntity {
    private readonly dv: Vector3Df;
    private readonly mv: number;
    public constructor();
    public constructor(d: Vector3Df, m: number);
    public constructor(o: Quaternionf | Quaterniond);
    public constructor(a: Vector3Df | Quaternionf | Quaterniond = new Vector3Df(), m = 0) {
        super();
        if (a instanceof Quaternionf) {
            this.dv = new Vector3Df(a.dv);
            this.mv = a.mv;
        } else if (a instanceof Quaterniond) {
            this.dv = new Vector3Df(a.direction());
            this.mv = f(a.magnitude());
        } else {
            if (a === null) throw new TypeError("Quaternion direction cannot be null");
            this.dv = new Vector3Df(a);
            this.mv = f(m);
        }
    }
    public static copyOf(o: Quaternionf): Quaternionf {
        if (o === null) throw new TypeError("Quaternion to copy cannot be null");
        return new Quaternionf(o);
    }
    public lengthSquared(): number {
        return f(f(this.mv * this.mv) + this.dv.dotProduct(this.dv));
    }
    public length(): number {
        return f(Math.sqrt(this.lengthSquared()));
    }
    public normalized(): Quaternionf {
        const l = this.length();
        return Math.abs(l) < f(VSDK.EPSILON) ? this : new Quaternionf(this.dv.multiply(f(1 / l)), f(this.mv / l));
    }
    public conjugated(): Quaternionf {
        return new Quaternionf(this.dv.multiply(-1), this.mv);
    }
    public rotate(v: Vector3Df): Vector3Df {
        if (v === null) throw new TypeError("Vector to rotate cannot be null");
        const uv = this.dv.crossProduct(v);
        return v.add(uv.multiply(f(2 * this.mv))).add(this.dv.crossProduct(uv).multiply(2));
    }
    public withDirection(d: Vector3Df): Quaternionf {
        return new Quaternionf(d, this.mv);
    }
    public withMagnitude(m: number): Quaternionf {
        return new Quaternionf(this.dv, m);
    }
    public direction(): Vector3Df {
        return this.dv;
    }
    public magnitude(): number {
        return this.mv;
    }
    public epsilonEquals(o: Quaternionf | null, e = f(VSDK.EPSILON)): boolean {
        if (o === null) return false;
        if (e < 0) throw new RangeError("epsilon must be >= 0");
        return this.dv.epsilonEquals(o.dv, e) && Math.abs(this.mv - o.mv) <= e;
    }
    public override toString(): string {
        return `${this.dv.toString()} / ${VSDK.formatDouble(this.mv)}`;
    }
}
